core/src/main/scala/schema/provider/metadata/SchemaConversions.scala

```diff
@@ -91,110 +98,129 @@ trait SchemaConversions {
+        case Name.INET      => M.pure(DataType.Inet)
+        case Name.INT       => M.pure(DataType.Int)
+        case Name.SMALLINT  => M.pure(DataType.Smallint)
+        case Name.TEXT      => M.pure(DataType.Text)
+        case Name.TIME      => M.pure(DataType.Time)
+        case Name.TIMESTAMP => M.pure(DataType.Timestamp)
+        case Name.TIMEUUID  => M.pure(DataType.Timeuuid)
+        case Name.TINYINT   => M.pure(DataType.Tinyint)
+        case Name.UUID      => M.pure(DataType.Uuid)
+        case Name.VARCHAR   => M.pure(DataType.Varchar)
+        case Name.VARINT    => M.pure(DataType.Varint)
         case _ =>
-          Left(SchemaDefinitionProviderError(s"Native DataType ${dataType.getName} not supported"))
+          M.raiseError(
+            SchemaDefinitionProviderError(s"Native DataType ${dataType.getName} not supported"))
       }
 
-    def toCollectionType(collectionType: CollectionType): SchemaResult[DataType] = {
+    def toCollectionType(collectionType: CollectionType): M[DataType] = {
 
       val typeArgs: List[DatastaxDataType] = collectionType.getTypeArguments.asScala.toList
 
-      val maybeCol = collectionType.getName match {
+      val maybeCol: Option[M[DataType]] = collectionType.getName match {
         case Name.LIST =>
           typeArgs.headOption map { typeArg =>
-            toDataTypeNative(typeArg) map DataType.List
+            M.map(toDataTypeNative(typeArg))(DataType.List)
           }
         case Name.SET =>
           typeArgs.headOption map { typeArg =>
-            toDataTypeNative(typeArg) map DataType.Set
+            M.map(toDataTypeNative(typeArg))(DataType.Set)
           }
         case Name.MAP =>
           for {
             t1 <- typeArgs.headOption
             t2 <- typeArgs.tail.headOption
-          } yield (toDataTypeNative(t1) |@| toDataTypeNative(t2)).map(DataType.Map)
+          } yield
+            M.map2(toDataTypeNative(t1), toDataTypeNative(t2))((t1, t2) => DataType.Map(t1, t2))
         case _ => None
       }
 
       maybeCol getOrElse {
-        Left(
+        M.raiseError(
           SchemaDefinitionProviderError(
             s"Error parsing collection DataType '${collectionType.asFunctionParameterString()}'"))
       }
     }
 
-    def toTupleType(tupleType: TupleType): SchemaResult[DataType] =
-      tupleType.getComponentTypes.asScala.toList traverse toDataTypeNative map DataType.Tuple
+    def toCustomType(className: String): M[DataType] =
+      M.pure(DataType.Custom(className))
+
+    def toUserDefinedType(keyspace: String, typeName: String): M[DataType] =
+      M.pure(DataType.UserDefined(KeyspaceName(keyspace), typeName))
+
+    def toTupleType(tupleType: TupleType): M[DataType] =
+      M.map(tupleType.getComponentTypes.asScala.toList.traverse(toDataTypeNative))(DataType.Tuple)
+
+    def toDataTypeNative2(dataType: DatastaxDataType): M[DataType] =
+      M.flatMap(toDataTypeNative(dataType))(M.pure(_))
```

Comment:
`M.flatMap(x)(M.pure(_))` is equal to just `x` (this is actually the right identity law for monad).

Reply:
Yeah, I'm doing this for casting the type, `DataType.Native` is a subtype of `DataType` but when I use the `toDataTypeNative` it gives me a compilation error (`expected DataType got DataType.Native`)

Reply:
You could do `M.widen[DataType](toDataTypeNative(dataType))` I think. (`widen` is defined on `Functor`).